fendermint/vm/interpreter/src/chain.rs

```diff
@@ -91,30 +103,52 @@ where
 
         // Append at the end - if we run out of block space, these are going to be reproposed in the next block.
         msgs.extend(ckpts);
+
+        // Prepare top down proposals
+        if let Some(proposal) = atomically(|| finality_provider.next_proposal()).await {
+            msgs.push(ChainMessage::Ipc(IpcMessage::TopDownExec(ParentFinality {
+                height: proposal.height as ChainEpoch,
+                block_hash: proposal.block_hash,
+            })))
+        }
+
         Ok(msgs)
     }
 
     /// Perform finality checks on top-down transactions and availability checks on bottom-up transactions.
-    async fn process(&self, pool: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
+    async fn process(&self, state: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
```

Comment:
```suggestion
    async fn process(&self, (pool, provider): Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
```
More readable than `state.0.get_status`